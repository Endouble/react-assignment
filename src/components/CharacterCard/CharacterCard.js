import React from 'react';
import PropTypes from 'prop-types';

import { Card, Button } from 'semantic-ui-react';

const CharacterCard = (props) => {
    const { character } = props;
    const showCharacterDetails = () => props.showCharacterModal(character);
    return (
        <Card fluid tabIndex='0' onKeyPress={showCharacterDetails}>
            <Card.Content>
                <Card.Header>
                    { character.name }
                </Card.Header>
            </Card.Content>
            <Card.Content extra>
                <Button onClick={showCharacterDetails} role="link" tabIndex="0"> View More </Button>
            </Card.Content>
        </Card>
    );
};

CharacterCard.propTypes = {
    character: PropTypes.shape({
        name: PropTypes.string.isRequired,
    }).isRequired,
    showCharacterModal: PropTypes.func.isRequired,
};

export default CharacterCard;
