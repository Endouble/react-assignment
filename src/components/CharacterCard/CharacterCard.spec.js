import React from 'react';
import { shallow, mount } from 'enzyme';

import CharacterCard from './CharacterCard';

describe('Character Card Test', () => {
  let props;
  let shallowCharacterCard;

  const getCharacterCard = () => {
    if (!shallowCharacterCard) {
      shallowCharacterCard = shallow(<CharacterCard {...props} />);
    }
    return shallowCharacterCard;
  };

  beforeEach(() => {
    props = {
      character: {
        name: "Luke Skywalker"
      },
      showCharacterModal: jest.fn()
    };

    shallowCharacterCard = undefined;
  });

  test('Component should render without crashing', () => {
    expect(getCharacterCard()).toBeDefined();
  });

  test('Component should render correctly', () => {
    expect(getCharacterCard()).toMatchSnapshot();
  });

  test('Component should render name', () => {
    const wrapper = mount(<CharacterCard {...props} />);
    const characterCard = wrapper.find('Card');
    const expected = expect.stringMatching(/^.*Luke Skywalker/);
    expect(characterCard.find('CardHeader').text()).toEqual(expected);
  });

  test('Component should render View More link', () => {
    const wrapper = mount(<CharacterCard {...props} />);
    const characterCard = wrapper.find('Card');
    const expected = expect.stringMatching(/^.*View More/);
    expect(characterCard.find('CardContent[extra=true]').text()).toEqual(expected);
  });

  test('Component should call showCharacterModal func on Click over View More link ', () => {
    const wrapper = getCharacterCard();
    const card = wrapper.find('Card CardContent[extra=true] Button');
    card.simulate('click', { preventDefault: jest.fn() });
    expect(props.showCharacterModal).toHaveBeenCalled();
  });

  test('Component should call showCharacterModal func on key press over card ', () => {
    const wrapper = getCharacterCard();
    const card = wrapper.find('Card');
    card.simulate('keypress', { preventDefault: jest.fn(), key: 'Enter' });
    expect(props.showCharacterModal).toHaveBeenCalled();
  });
})