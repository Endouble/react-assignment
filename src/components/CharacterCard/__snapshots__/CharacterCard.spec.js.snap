// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Character Card Test Component should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CharacterCard
    character={
        Object {
            "name": "Luke Skywalker",
          }
    }
    showCharacterModal={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <CardContent>
          <CardHeader>
                    Luke Skywalker
          </CardHeader>
</CardContent>,
        <CardContent
          extra={true}
>
          <Button
                    as="button"
                    onClick={[Function]}
                    role="link"
                    tabIndex="0"
          >
                     View More 
          </Button>
</CardContent>,
      ],
      "fluid": true,
      "onKeyPress": [Function],
      "tabIndex": "0",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <CardHeader>
            Luke Skywalker
</CardHeader>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Luke Skywalker",
          },
          "ref": null,
          "rendered": "Luke Skywalker",
          "type": [Function],
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <Button
            as="button"
            onClick={[Function]}
            role="link"
            tabIndex="0"
>
             View More 
</Button>,
          "extra": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "as": "button",
            "children": " View More ",
            "onClick": [Function],
            "role": "link",
            "tabIndex": "0",
          },
          "ref": null,
          "rendered": " View More ",
          "type": [Function],
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <CardContent>
            <CardHeader>
                        Luke Skywalker
            </CardHeader>
</CardContent>,
          <CardContent
            extra={true}
>
            <Button
                        as="button"
                        onClick={[Function]}
                        role="link"
                        tabIndex="0"
            >
                         View More 
            </Button>
</CardContent>,
        ],
        "fluid": true,
        "onKeyPress": [Function],
        "tabIndex": "0",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <CardHeader>
              Luke Skywalker
</CardHeader>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Luke Skywalker",
            },
            "ref": null,
            "rendered": "Luke Skywalker",
            "type": [Function],
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <Button
              as="button"
              onClick={[Function]}
              role="link"
              tabIndex="0"
>
               View More 
</Button>,
            "extra": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "as": "button",
              "children": " View More ",
              "onClick": [Function],
              "role": "link",
              "tabIndex": "0",
            },
            "ref": null,
            "rendered": " View More ",
            "type": [Function],
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
